sdk/python/kfp/components/structures.py

```diff
@@ -119,13 +116,21 @@ def __eq__(self, other: Any) -> bool:
         """
         if isinstance(other, InputSpec):
             return type_utils.get_canonical_name_for_outer_generic(
-                self.type
-            ) == type_utils.get_canonical_name_for_outer_generic(
-                other.type
-            ) and self.default == other.default and self.schema_version == other.schema_version
+                self.type) == type_utils.get_canonical_name_for_outer_generic(
+                    other.type) and self.default == other.default
         else:
             return False
 
+    def validate_type(self) -> None:
+        """Type should either be a parameter or a valid bundled artifact type
+        by the time it gets to InputSpec.
+
+        This allows us to perform fewer checks downstream.
+        """
+        # TODO: add transformation logic so that we don't have to transform inputs at every place they are used, including v1 back compat support
+        if not spec_type_is_parameter(self.type):
+            type_utils.validate_bundled_artifact_type(self.type)
+
```

Comment:
This validation allows us to drastically simplify much of the compilation logic and inputspec/outputspec checking we do to construct IR.

TODO for a future refactor PR.

Reply:
I'm not sure about coupling that introduces dependency from the structure types in `structures.py` to other part of the library. There was some additional validation and processing logic in few of the classes, but I'm now thinking some of that might have been a mistake (e.g. `_toposorted_tasks`).